### ISSUE TITLE
new URL is unable to parse 'strange' protocols

### ISSUE THREAD

**Issue by aricart** [denoland/deno#5410]
Unable to parse a non-http URL like 'nats://localhost:4222', same form with an `https` protocol works fine.
```
Deno.test({
    name: 'url',
    fn(): void {
        let u = new URL("nats://localhost:4222")
        assertEquals('localhost', u.hostname)
        assertEquals('4222', u.port)
    }
})
```
Associated with this issue, the lack of insight into an unexpected value. In my code, at some point:

```
            const u = new URL(options.url)
            const hp = {hostname: u.hostname, port: parseInt(u.port, 10)}
            console.log(hp)
            Deno.connect(hp)
...
```

The error at runtime is completely unhelpful:
```
InvalidData: data did not match any variant of untagged enum ArgsEnum
    at unwrapResponse ($deno$/ops/dispatch_json.ts:43:11)
    at Object.sendAsync ($deno$/ops/dispatch_json.ts:98:10)
    at async Object.connect ($deno$/net.ts:180:11)
```

**Comment by kitsonk** [comment]
@aricart this appears to match the behaviour of `URL` in the browser.  In Chrome:

```js
const u = new URL("nats://localhost:4222");
u.hostname; // ""
u.port; // ""
```

This is because the specification says that only certain schemes, is the [host considered be the domain name](https://url.spec.whatwg.org/#concept-url-host).

**Comment by aricart** [comment]
Yes, I get that - the browser will do what you are doing. With that said, all languages are able to provide some information, wondering if instead, deno should provide an URI type class, to match the Java guys.

On my side, I can work around it - by changing the protocol.



```python
from urllib.parse import urlsplit
o = urlsplit('nats://localhost:4222')
print(o)
```
> SplitResult(scheme='nats', netloc='localhost:4222', path='', query='', fragment='')

```Go
import (
    "net/url"
    "fmt"
)
u, _ := url.Parse("nats://localhost:4222")
fmt.Println(u.Host)
```
> localhost:4222

```Javascript
const url = require('url')
const u = new url.URL('nats://localhost:4222')
console.log(u.hostname, u.port, u.host, u.protocol)
```
> localhost 4222 localhost:4222 nats:

```java
import java.net.*;
import java.io.*;

URI u = new URI("nats://localhost:4222");
System.out.println("protocol: " + u.getScheme() + " host: " + u.getHost() + " port: " + u.getPort());
```
> protocol: nats host: localhost port: 4222

**Comment by kitsonk** [comment]
> With that said, all languages are able to provide some information, wondering if instead, deno should provide an URI type class, to match .... Java ...

We have been following the standards.  Having a redundant class built in seems illogical.  Having something in `std` that is an opt in alternative, maybe.

**Comment by aricart** [comment]
URI would be fine - for now, it is easy to work around it. Other features are more important.

**Comment by wschenk** [comment]
I'm not sure I understand which standard you are following -- I guess its more of a "deno is a command line browser" thing -- but I found this behavior unexpected.  I'm working around it with regex, which is always a bit weird...

```ts
const redis_url = new URL( REDIS_URL.replace( /^redis:/, "http:" ) )
```

**Comment by kitsonk** [comment]
@wschenk https://url.spec.whatwg.org/#url-miscellaneous lists special schemes.  Then https://url.spec.whatwg.org/#url-representation specifies what schemes get parsed in what way.  Only special schemes (excluding `file`) get parsed out that way.

**Comment by wschenk** [comment]
Thanks @kitsonk that is super interesting.  What happened to gopher://?  Living standards I guess.

**Comment by bartlomieju** [comment]
@nayeemrmn has this issue been fixed in #6689?

**Comment by nayeemrmn** [comment]
I think the specified behaviour is that it _should_ parse a host for non-special protocols using the [opaque-host parser](https://url.spec.whatwg.org/#concept-opaque-host-parser). Browsers aren't the most compliant or consistent with each other when it comes to `URL`. See https://jsdom.github.io/whatwg-url/#url=bmF0czovL2xvY2FsaG9zdDo0MjAwLw==&base=ZmlsZTovLy8= instead.

This was fixed in #6689... but I left a bug where it doesn't parse a port for non-special protocols 😛 Will link this issue in the fix.
